blscurve/bls_batch_verifier.nim

```diff
@@ -0,0 +1,376 @@
+    # as there is more work than hardware threads/batches
+    # and it's better to accumulate as much as possible in partial pairings
+    # as merge cost is non-trivial (10k to 20k cycles).
+    # So we want 1 merge per thread and no more.
+    let ok = accumPairingLines(sigsets, contexts,
+                               batchID, subsetStart, subsetStopEx)
+    if not ok:
+      return false
+    return true
+  elif batchID.right >= numBatches:
+    # No need to split, there is no right subtree
+    # We directly accumulate in the current batch context
+    let ok = accumPairingLines(sigsets, contexts,
+                               batchID, subsetStart, mid)
+    if not ok:
+      return false
+    return true
+  else:
+    # Split in half, distribute the right and work on the left
+    var leftOk, rightOk = true
+    omp_task"":
+      rightOk = dacPairing(sigsets, contexts, numBatches,
+                           batchID.right(), mid, subsetStopEx)
+    leftOk = dacPairing(sigsets, contexts, numBatches,
+                        batchID.left(), subsetStart, mid)
+
+    # Wait for all subtrees
+    omp_taskwait()
+    if not leftOk or not rightOk:
+      return false
+    # Merge left and right
+    let mergeOk = contexts[batchID.left()].merge(contexts[batchID.right()])
+    if not mergeOk:
+      return false
+    # Update our own pairing context - 3+MB copy
+    contexts[batchID] = contexts[batchID.left()]
+    return true
+
+{.pop.}
+
+proc batchVerifyParallel*(
+       batcher: var BatchedBLSVerifier,
+       secureRandomBytes: array[32, byte]
+     ): bool {.sideeffect.} =
+  ## Multithreaded batch verification
+  ## Requires OpenMP 3.0 (GCC 4.4, 2008)
+  let numSets = batcher.sets.len.uint32
+  if numSets == 0:
+    return false
+
+  let numBatches = min(numSets, omp_get_num_threads().uint32)
+
+  # Stage 0: Accumulators - setLen for noinit of seq
+  batcher.batchContexts.setLen(numBatches.int)
+
+  # No stacktrace, exception
+  # or anything that require a GC in a parallel section
+  # otherwise "attachGC()" is needed in the parallel prologue
+  # Hence we use raw ptr UncheckedArray instead of seq
+  let contextsPtr = batcher.batchContexts.toPtrUncheckedArray()
+  let setsPtr = batcher.sets.toPtrUncheckedArray()
+
+  var ok: bool
+  {.push stacktrace:off, checks:off.}
+  omp_parallel:
+    let threadID = omp_get_thread_num()
+
+    contextsPtr[threadID].init(
+      secureRandomBytes,
+      threadSepTag = cast[array[sizeof(cint), byte]](threadID)
+    )
+
+    omp_single:
+      ok = dacPairing(setsPtr, contextsPtr, numBatches,
+                      batchID = 0, subsetStart = 0, subsetStopEx = numSets)
+  {.pop.}
+
+  if not ok:
+    return false
+
+  return batcher.batchContexts[0].finalVerify()
+
+# Autoselect Batch Verifier
+# ----------------------------------------------------------------------
+
+proc batchVerify*(
+       batcher: var BatchedBLSVerifier,
+       secureRandomBytes: array[32, byte]
```

Comment:
how often do the bytes need to be updated? every call?

Reply:
Unfortunately that's a case of rolling my own crypto but here are thoughts that weigh in my decision to have a new one at each batchVerify:

- What we want is something in the (PublicKey, message, Signature) triplet that is not known or forgeable by the attacker.
- Due to bilinearity of pairings if (PK1, MSG1, S1) and (PK2, MSG2, S2) verifies we have:
  - (r*PK1, MSG1, r*S1) that verifies (r being a random scalar)
  - (PK1, r*MSG1, r*S1) that verifies
  - (PK1+PK2, MSG1+MSG2, S1+S2) that verifies (+ being an aggregation function)
  - (PK1+FakePK+PK2-FakePK, MSG1+MSG2, S1+S2) that verifies
- Because PK1 and PK2 are public anyone could craft keys (PK1+FakePK, PK2-FakePK) that would verify S1+S2 but individually would not verify (S1, S2).
- This is what the random r factor defends against.
- r * S1 is very costly if r is the maximum size (25% of a full pairing, would make batch pairing not much of an improvement)
- Instead we could use only 64 bits and reduce the overhead to below 10%. This assumes that an attacker cannot retry 2^64 times with forged public keys on the same message and signature (or maybe 2^32 if the birthday paradox applies).
- So we need to change the seed at least once every 2^32 calls.
- For simplicity, and in case new advances arise, we can change that every call and relax that later if it proves a bottleneck
- Also we don't defend private data so there is no stringent requirement on the RNG
- Hence instead of requiring a callback to a RNG that we would call every time we include a new (PK, msg, Sig) triplets
  or requiring passing (PK, msg, Sig, randomBytes), we cache a random byte seed at init and derive new random numbers at each update with SHA256 to avoid depleting the local machine entropy pool, costly randomization at the kernel level and impact other critical crypto primitives (for example random peer selection).

Reply:
ok - let's make sure this rationale is in a comment in the code